sources/api/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,315 @@
+            // this condition is already taken care of
+            let (cmd, args) = command.get_args();
+            let arg_str: Vec<&str> = args.iter().map(String::as_str).collect();
+            run_command(cmd.as_str(), arg_str)?;
+        }
+
+        if mode == "once" {
+            let formatted = format!("settings.bootstrap-commands.{}.mode=off", name);
+            info!("Turning off bootstrap-command '{}'", name);
+            run_command("apiclient", ["set", formatted.as_str()])?;
+        }
+
+        info!("Succesfully ran bootstrap command '{}'", name);
+    }
+
+    Ok(())
+}
+
+/// Query the API for the currently defined bootstrap commands
+fn get_bootstrap_commands<P>(config_path: P) -> Result<BTreeMap<Identifier, BootstrapCommand>>
+where
+    P: AsRef<Path>,
+{
+    let config_str = fs::read_to_string(config_path.as_ref()).context(error::ReadConfigSnafu {
+        config_path: config_path.as_ref(),
+    })?;
+
+    let config: BootstrapCommandConfig =
+        toml::from_str(&config_str).context(error::DeserializationSnafu {
+            config_path: config_path.as_ref(),
+        })?;
+
+    Ok(config.bootstrap_commands.unwrap_or_default())
+}
+
+/// Parse the args to the program and return an Args struct
+fn parse_args(args: env::Args) -> Result<Args> {
+    let mut global_args = Args::default();
+
+    let mut iter = args.skip(1);
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            // Global args
+            "--log-level" => {
+                let log_level = iter.next().context(error::UsageSnafu {
+                    message: "Did not give argument to --log-level",
+                })?;
+                global_args.log_level = LevelFilter::from_str(&log_level)
+                    .context(error::LogLevelSnafu { log_level })?;
+            }
+
+            "-c" | "--config-path" => {
+                let config_str = iter.next().context(error::UsageSnafu {
+                    message: "Did not give argument to --config-path",
+                })?;
+                global_args.config_path = PathBuf::from(config_str.as_str());
+            }
+
+            _ => (),
+        }
+    }
+
+    Ok(global_args)
+}
+
+fn run() -> Result<()> {
+    let args = parse_args(env::args())?;
+
+    // SimpleLogger will send errors to stderr and anything less to stdout.
+    SimpleLogger::init(args.log_level, LogConfig::default()).context(error::LoggerSnafu)?;
+
+    let bootstrap_commands = get_bootstrap_commands(args.config_path)?;
+
+    // We want to run bootstrap commands in sorted order of their names.
+    for (name, bootstrap_command) in bootstrap_commands.iter() {
+        let essential = bootstrap_command.essential.unwrap_or(false);
+        let status = handle_bootstrap_command(name, bootstrap_command);
+
+        ensure!(
+            !essential || status.is_ok(),
+            error::BootstrapCommandFailureSnafu {
+                name: name.as_ref()
+            }
+        )
```

Comment:
This intersects with the "once" behavior in a strange way: if the first set of commands is marked as "essential" and "once", and it fails, then the boot should fail.

But, we have to decide what to do on a subsequent boot:
1. We already ran the command once, so we don't have to run it again, even though it wasn't successful. This exposes later commands in the list to an unexpected state - maybe they expect previous commands to have all been executed successfully, but on the next boot we'll skip over the failed ones.
2. We ran the command already, but it failed, so we need to run it again. This implies "once" means "one successful execution". This is easier for later commands in the list to reason about - they can be sure that previous essential commands were all executed successfully.

I lean toward (2) as it's consistent with how bootstrap containers are handled; we don't disable a "once" container unless it returns success.

Relative to my earlier comment, you can implement this by creating `/var/lib/bottlerocket/bootstrap-commands@<name>.ran` after the command returns success, rather than before.

Reply:
Resolved.